File: Objective C/testTek/SFTableViewCell.m
//
//  SFTableViewCell.m
//  testTek
//
//  Created by applecaca on 09/09/15.
//  Copyright © 2015 stupeflix. All rights reserved.
//

#import "SFTableViewCell.h"
#import "DataController.h"
#import "SFTableViewLayout.h"

const NSUInteger    kCandidateModeOffMargin = 10;
const NSUInteger    kCandidateModeOnMargin = 115;
const CGFloat       kCellHeight            = 95;

@interface SFTableViewCell()
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;

@property (nonatomic, strong) NSArray *constraints;
@property (nonatomic, strong) NSLayoutConstraint *topConstraint;
@property (nonatomic, strong) NSLayoutConstraint *botomConstraint;

@property (nonatomic, weak) SFTableViewCell *previousCell;
@property (nonatomic, weak) SFTableViewCell *nextCell;

- (void)activateCandidateMode;
- (void)desactivateCandidateMode;

@end


@implementation SFTableViewCell


/*  Methode de classe qui alloue une instance de SFTableViewCell à partir du xib puis set le modele Employe */
+ (SFTableViewCell *)cellWithEmploye:(Employe *)pEmploye
{
    SFTableViewCell *cell = [[NSBundle mainBundle] loadNibNamed:@"SFTableViewCell" owner:nil options:nil].firstObject;
    [cell configureWithEmploye:pEmploye];
    [cell setTranslatesAutoresizingMaskIntoConstraints:NO];
    return cell;
}


/*  Methode de classe qui renvoie la hauteur d'une cellule utile pour la scrollview avec autolayout */
+ (CGFloat)height
{
    return kCellHeight;
}


/*  Methode de classe qui renvoie la largeur d'une cellule utile pour la scrollview avec autolayout */
+ (CGFloat)width
{
    return [UIScreen mainScreen].bounds.size.width;
}


/* Configure la cellule avec le modèle Employe */
- (void)configureWithEmploye:(Employe *)pEmploye
{
    self.employe = pEmploye;
    if (pEmploye == nil)
    {
        self.name.text = nil;
        self.job.text = nil;
        self.email.text = nil;
        self.spinner.hidden = YES;
        self.imageView.image = nil;
        return;
    }
    self.name.text = pEmploye.name;
    self.job.text = pEmploye.job;
    self.email.text = pEmploye.email;
    
    self.spinner.hidden = NO;
    [self.spinner startAnimating];
    [[DataController sharedInstance] fetchImageWithURL:pEmploye.picture AndCompletionBlock:^(UIImage *image, NSError *error)
     {
         dispatch_async(dispatch_get_main_queue(), ^
         {
             self.imageView.image = image;
             [self.spinner stopAnimating];
             [self.spinner setHidden:YES];
         });
         
         
     }];
}


/*  Cette méthode permet de set un Dictionnaire qui contient toute les contraintes 
    nécessaire au placement de la cellule (kAllConstraints) mais egallement des pointeurs vers
    les contraintes top et botom car leur constantes seront modifier pour le mode candiate (decallage)
 */
- (void)setConstraintsDictionnary:(NSDictionary *)pDicoContraints
{
    if (pDicoContraints == nil || ! [pDicoContraints isKindOfClass:[NSDictionary class]]) return;
    
    self.constraints = pDicoContraints[kAllConstraints];
    self.topConstraint = pDicoContraints[kTopConstraint];
    self.botomConstraint = pDicoContraints[kBottomConstraint];
}


/*  Active le mode candidate (decallage)
    la constante de la contrainte top prend la valeur 115 soit la taille de la cellule + 2 fois la marge (10x2)
 */
- (void)activateCandidateMode
{
    self.topConstraint.constant = kCandidateModeOnMargin;
}


/*  Desactive le mode candidate (decallage)
    la constante de la contrainte top ou botom reprend sa valeur initiale 10 ou 0 
    si c'est une cellule de bord
 */
- (void)desactivateCandidateMode
{
    self.backgroundColor = [UIColor whiteColor];
    
    if (self.previousCell == nil)
    {
        self.topConstraint.constant = 0;
    }
    else
    {
        self.topConstraint.constant = kCandidateModeOffMargin;
    }
    
    if (self.nextCell == nil)
    {
        self.botomConstraint = 0;
    }
}



@end

File: Swift/testTekSwift/testTekSwift/SFTableViewCell.swift
//
//  SFTableViewCell.swift
//  testTekSwift
//
//  Created by Nicolas Bellon on 28/10/2015.
//  Copyright © 2015 daid. All rights reserved.
//

import Foundation
import UIKit


let kCandidateModeOffMargin: CGFloat = 10;
let kCandidateModeOnMargin: CGFloat = 140;

class SFTableViewCell: UIView
{
    @IBOutlet weak var spinner: UIActivityIndicatorView!
    @IBOutlet weak var name: UILabel!
    @IBOutlet weak var job: UILabel!
    @IBOutlet weak var email: UILabel!
    @IBOutlet weak var picture: UIImageView!
    
    var previousCell : SFTableViewCell?;
    var nextCell : SFTableViewCell?;
    var cellConstraint: [NSLayoutConstraint] = [];
    var topConstraint: NSLayoutConstraint?

    private func setupEmploye(employe: Employe)
    {
        self.name.text = employe.name;
        self.job.text = employe.job;
        self.email.text = employe.email;
        
        self.spinner.hidden = false;
        self.spinner.startAnimating();
        
        DataController.sharedInstance.fetchImageWithUrl(employe.picture) { (image: UIImage?, error: NSError?) -> Void in
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                self.picture.image = image;
                self.spinner.stopAnimating();
                self.spinner.hidden = true;
            })
        }
    }
    
    class func initWithEmploye(employe: Employe) -> SFTableViewCell?
    {
        let dest : SFTableViewCell? = NSBundle.mainBundle().loadNibNamed("SFTableViewCell", owner: nil, options: nil).first as! SFTableViewCell?;

        if dest == nil
        {
            return nil;
        }
        
        dest!.setupEmploye(employe);
        return dest;
    }
    
    class func cellHeight() -> Float
    {
        return 120.0;
    }
    
    class func cellWidth() -> Float
    {
        return Float(UIScreen.mainScreen().bounds.width);
    }

    
    override init(frame: CGRect)
    {
        super.init(frame: frame);
    }

    required init?(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder);
    }
    
    func desactivateCandidateMode()
    {
        guard self.topConstraint != nil
        else
        {
            return;
        }
        
        if self.previousCell == nil
        {
            self.topConstraint!.constant = 0;
        }
        else
        {
            self.topConstraint!.constant = kCandidateModeOffMargin;
        }
    }
    
    func activateCandidateMode()
    {
        guard let topConstraint = self.topConstraint
        else
        {
            return;
        }
        
        topConstraint.constant = kCandidateModeOnMargin;
    }
}